crates/cext/src/dag.rs

```diff
@@ -736,3 +736,62 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
         }
     }
 }
+/// @ingroup QkDag
+/// Return the operation nodes in the DAG listed in topological order.
+///
+/// @param dag A pointer to the DAG.
+/// @param order A pointer to an array of ``qk_dag_num_op_nodes(dag)`` elements
+/// of type ``uint32_t``, where this function will write the output to.
+///
+/// # Example
+/// ```c
+/// QkDag *dag = qk_dag_new();
+/// QkQuantumRegister *qr = qk_quantum_register_new(1, "my_register");
+/// qk_dag_add_quantum_register(dag, qr);
+///
+/// uint32_t qubit[1] = {0};
+/// qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+/// qk_dag_apply_gate(dag, QkGate_S, qubit, NULL, false);
+///
+/// // get the number of operation nodes
+/// uint32_t num_ops = qk_dag_num_op_nodes(dag); // 2
+/// uint32_t *order = malloc(sizeof(uint32_t) * num_ops);
+///
+/// // get operation nodes listed in topological order
+/// qk_dag_topological_op_nodes(dag, order);
+///
+/// // do something with the ordered nodes
+/// for (uint32_t i=0; i<num_ops; i++) {
+///     QkGate gate = qk_dag_op_node_gate_op(dag, order[i], NULL);
+///     printf("The gate at location %u is %u.\n", i, gate);
+/// }
+///
+/// // free the order array
+/// free(order);
+/// qk_quantum_register_free(qr);
+/// qk_dag_free(dag);
+/// ```
+///
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not a valid, non-null pointer to a ``QkDag``
+/// or if ``order`` is not a valid, non-null pointer to a sequence of ``qk_dag_num_op_nodes(dag)``
+/// consecutive elements of ``uint32_t``.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, order: *mut u32) {
+    // SAFETY: Per documentation, ``dag`` is non-null and valid.
+    let dag = unsafe { const_ptr_as_ref(dag) };
+
+    let out_topological_op_nodes = dag.topological_op_nodes().unwrap();
+
+    // SAFETY: Per documentation, ``order`` is a valid pointer with a sufficient allocation for the output
+    // array.
+    unsafe {
+        let out_slice = std::slice::from_raw_parts_mut(order, dag.num_ops());
+        out_slice
+            .iter_mut()
+            .zip(out_topological_op_nodes)
+            .for_each(|(dest, src)| *dest = src.index() as u32);
+    };
```

Comment:
```suggestion
    let out_slice = unsafe { std::slice::from_raw_parts_mut(out_order, dag.num_ops()) };
    out_slice
        .iter_mut()
        .zip(out_topological_op_nodes)
        .for_each(|(dest, src)| *dest = src.index() as u32);
```
It's more idiomatic to minimize the scope of `unsafe`.